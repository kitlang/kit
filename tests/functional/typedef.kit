struct MyStruct {
    public static function staticMethod() {
        puts("MyStruct.staticMethod()");
    }

    public var a: Int;
    public var b: Int;

    public function instanceMethod() {
        puts("MyStruct.instanceMethod()");
    }
}

typedef MyTypedef = MyStruct;

struct MyGenericStruct[$N] {
    public static function getLength(): Int {
        return N;
    }

    var arr: CArray[Int, N] = empty;
}

typedef GenStruct4 = MyGenericStruct[4];
typedef GenStruct8 = MyGenericStruct[8];

function main() {
    MyTypedef.staticMethod();
    var x: MyTypedef = struct MyTypedef {a: 1, b: 2};
    x.instanceMethod();
    var y: MyStruct = struct MyStruct{a: 3, b: 4};
    var z = y as MyTypedef;

    var genStruct4 = struct GenStruct4;
    var genStruct8 = struct GenStruct8;
    printf("%i\n", GenStruct4.getLength());
    printf("%i\n", GenStruct8.getLength());
}
