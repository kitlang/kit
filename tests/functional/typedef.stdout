MyStruct.staticMethod()
MyStruct.instanceMethod()
4
8
